File: Examples/ios-test/LedgerBtcExample/BLETransport.swift
// CoreBluetooth implementation of ledger-btc-core's Transport
// protocol for the Ledger Nano X on macOS.
//
// Owned by this CLI example, not by the SDK. Production iOS / Mac
// hosts would write their own version; the protocol shape is what
// matters. This is the reference implementation for what each
// platform layer needs to do.

import CoreBluetooth
import Foundation

private let ledgerServiceUUID = CBUUID(string: "13d63400-2c97-0004-0000-4c6564676572")
private let ledgerWriteUUID   = CBUUID(string: "13d63400-2c97-0004-0002-4c6564676572")
private let ledgerNotifyUUID  = CBUUID(string: "13d63400-2c97-0004-0001-4c6564676572")

// Standard Battery Service for the keep-alive heartbeat.
private let batteryServiceUUID = CBUUID(string: "180F")
private let batteryLevelUUID   = CBUUID(string: "2A19")

// Per the Ledger BLE protocol doc: tag(0x05) || idx_be(2) ||
// [first packet only: total_len_be(2)] || chunk.
private let bleApduTag: UInt8 = 0x05

// Verified-stable safe MTU on Nano X. macOS sometimes reports
// 512 via maximumWriteValueLength but the device actually caps
// around 153.
private let safeMTU = 153

final class BLETransport: NSObject, @unchecked Sendable, Transport {
    private let verbose: Bool

    private var central: CBCentralManager?
    private var peripheral: CBPeripheral?
    private var writeChar: CBCharacteristic?
    private var notifyChar: CBCharacteristic?
    private var batteryLevelChar: CBCharacteristic?

    // One-shot continuations resolved by the CB delegates.
    private var poweredOnCont: CheckedContinuation<Void, Error>?
    private var connectedCont: CheckedContinuation<Void, Error>?
    private var discoveredCont: CheckedContinuation<Void, Error>?

    // The exchange()-in-flight reassembly state.
    private var pendingAPDU: PendingAPDU?
    private struct PendingAPDU {
        var totalLength: Int
        var buffer: Data
        var continuation: CheckedContinuation<Data, Error>?
    }

    init(verbose: Bool = false) {
        self.verbose = verbose
        super.init()
    }

    func connect() async throws {
        if let p = peripheral, p.state == .connected, writeChar != nil, notifyChar != nil {
            return
        }
        central?.stopScan()
        if let p = peripheral, let c = central {
            c.cancelPeripheralConnection(p)
        }
        peripheral = nil
        writeChar = nil
        notifyChar = nil
        batteryLevelChar = nil

        if central == nil {
            central = CBCentralManager(delegate: self, queue: nil)
        }
        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
            if let central, central.state == .poweredOn {
                cont.resume()
            } else {
                self.poweredOnCont = cont
            }
        }
        log("scan: looking for Ledger service")
        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
            self.connectedCont = cont
            self.central?.scanForPeripherals(withServices: [ledgerServiceUUID], options: nil)
        }
        log("connected, discovering services")
        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
            self.discoveredCont = cont
            self.peripheral?.discoverServices([ledgerServiceUUID, batteryServiceUUID])
        }
        log("ready: write=\(writeChar?.uuid.uuidString ?? "?") notify=\(notifyChar?.uuid.uuidString ?? "?")")
    }

    // MARK: -- Transport protocol

    func exchange(apdu: Data) async throws -> ExchangeResponse {
        guard let writeChar, let peripheral, peripheral.state == .connected else {
            throw TransportError.Disconnected(reason: "BLE not connected")
        }
        if let stale = pendingAPDU?.continuation {
            stale.resume(throwing: TransportError.Io(reason: "previous exchange abandoned"))
        }
        pendingAPDU = nil

        log(">>> APDU (\(apdu.count) B): \(apdu.hexEncodedString())")

        // Keep-alive heartbeat: 400ms initial delay, then 500ms
        // intervals. Verified stable across the 200+ round
        // SIGN_PSBT flow on macOS in the research-stage Swift port.
        let heartbeat = Task { @MainActor [weak self] in
            do {
                try await Task.sleep(nanoseconds: 400_000_000)
                while !Task.isCancelled {
                    guard let self,
                          self.peripheral?.state == .connected,
                          let batChar = self.batteryLevelChar else { break }
                    self.peripheral?.readValue(for: batChar)
                    try await Task.sleep(nanoseconds: 500_000_000)
                }
            } catch {
                // cancellation, parent op completed
            }
        }
        defer { heartbeat.cancel() }

        let raw = try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Data, Error>) in
            self.pendingAPDU = PendingAPDU(totalLength: -1, buffer: Data(), continuation: cont)
            let reported = peripheral.maximumWriteValueLength(for: .withResponse) - 3
            let mtu = max(20, min(safeMTU, reported))
            for packet in framedPackets(apdu: apdu, mtu: mtu) {
                peripheral.writeValue(packet, for: writeChar, type: .withResponse)
            }
        }
        guard raw.count >= 2 else {
            throw TransportError.Io(reason: "APDU response too short (\(raw.count) bytes)")
        }
        let sw = (UInt16(raw[raw.count - 2]) << 8) | UInt16(raw[raw.count - 1])
        let data = Data(raw.prefix(raw.count - 2))
        log("<<< sw=0x\(String(format: "%04X", sw)) data=\(data.count) B")
        return ExchangeResponse(statusWord: sw, data: data)
    }

    private func log(_ s: String) {
        guard verbose else { return }
        FileHandle.standardError.write("[\(ts())] \(s)\n".data(using: .utf8)!)
    }

    private func ts() -> String {
        let fmt = ISO8601DateFormatter()
        fmt.formatOptions = [.withInternetDateTime]
        return fmt.string(from: Date())
    }
}

// MARK: -- BLE framing helpers

private func framedPackets(apdu: Data, mtu: Int) -> [Data] {
    var packets: [Data] = []
    var remaining = apdu
    var idx: UInt16 = 0
    while !remaining.isEmpty {
        var header = Data()
        header.append(bleApduTag)
        header.append(UInt8(idx >> 8))
        header.append(UInt8(idx & 0xff))
        if idx == 0 {
            header.append(UInt8(apdu.count >> 8))
            header.append(UInt8(apdu.count & 0xff))
        }
        var take = mtu - header.count
        if take > remaining.count { take = remaining.count }
        var pkt = header
        pkt.append(remaining.prefix(take))
        packets.append(pkt)
        remaining = remaining.dropFirst(take)
        idx &+= 1
    }
    return packets
}

extension Data {
    fileprivate func hexEncodedString() -> String {
        map { String(format: "%02x", $0) }.joined()
    }
}

// MARK: -- CB delegates

extension BLETransport: CBCentralManagerDelegate {
    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        switch central.state {
        case .poweredOn:
            poweredOnCont?.resume()
            poweredOnCont = nil
        case .unauthorized:
            poweredOnCont?.resume(throwing: TransportError.Io(
                reason: "Bluetooth permission denied. Grant access in System Settings."
            ))
            poweredOnCont = nil
        case .poweredOff:
            poweredOnCont?.resume(throwing: TransportError.Io(
                reason: "Bluetooth is off. Enable Bluetooth and retry."
            ))
            poweredOnCont = nil
        case .unsupported:
            poweredOnCont?.resume(throwing: TransportError.Io(
                reason: "This Mac does not support Bluetooth LE."
            ))
            poweredOnCont = nil
        default:
            break
        }
    }

    func centralManager(_ central: CBCentralManager,
                        didDiscover peripheral: CBPeripheral,
                        advertisementData: [String: Any],
                        rssi RSSI: NSNumber) {
        log("discovered \(peripheral.identifier.uuidString) rssi=\(RSSI)")
        central.stopScan()
        self.peripheral = peripheral
        peripheral.delegate = self
        central.connect(peripheral, options: nil)
    }

    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        connectedCont?.resume()
        connectedCont = nil
    }

    func centralManager(_ central: CBCentralManager,
                        didFailToConnect peripheral: CBPeripheral,
                        error: Error?) {
        let reason = error?.localizedDescription ?? "unknown"
        connectedCont?.resume(throwing: TransportError.Disconnected(
            reason: "connect failed: \(reason)"
        ))
        connectedCont = nil
    }

    func centralManager(_ central: CBCentralManager,
                        didDisconnectPeripheral peripheral: CBPeripheral,
                        error: Error?) {
        let reason = error?.localizedDescription ?? "device disconnected"
        let err = TransportError.Disconnected(reason: reason)
        pendingAPDU?.continuation?.resume(throwing: err)
        pendingAPDU = nil
        connectedCont?.resume(throwing: err)
        connectedCont = nil
        discoveredCont?.resume(throwing: err)
        discoveredCont = nil
        writeChar = nil
        notifyChar = nil
        batteryLevelChar = nil
        self.peripheral = nil
    }
}

extension BLETransport: CBPeripheralDelegate {
    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        if let error {
            discoveredCont?.resume(throwing: TransportError.Disconnected(
                reason: "service discovery failed: \(error.localizedDescription)"
            ))
            discoveredCont = nil
            return
        }
        for svc in peripheral.services ?? [] {
            if svc.uuid == ledgerServiceUUID {
                peripheral.discoverCharacteristics([ledgerWriteUUID, ledgerNotifyUUID], for: svc)
            } else if svc.uuid == batteryServiceUUID {
                peripheral.discoverCharacteristics([batteryLevelUUID], for: svc)
            }
        }
    }

    func peripheral(_ peripheral: CBPeripheral,
                    didDiscoverCharacteristicsFor service: CBService,
                    error: Error?) {
        let chars = service.characteristics ?? []
        if service.uuid == batteryServiceUUID {
            batteryLevelChar = chars.first { $0.uuid == batteryLevelUUID }
            return
        }
        writeChar = chars.first { $0.uuid == ledgerWriteUUID }
        notifyChar = chars.first { $0.uuid == ledgerNotifyUUID }
        guard let notifyChar else {
            discoveredCont?.resume(throwing: TransportError.Io(
                reason: "Ledger notify characteristic missing"
            ))
            discoveredCont = nil
            return
        }
        peripheral.setNotifyValue(true, for: notifyChar)
        discoveredCont?.resume()
        discoveredCont = nil
    }

    func peripheral(_ peripheral: CBPeripheral,
                    didUpdateValueFor characteristic: CBCharacteristic,
                    error: Error?) {
        if characteristic.uuid == batteryLevelUUID { return }
        guard characteristic.uuid == ledgerNotifyUUID, let data = characteristic.value else { return }
        guard var pending = pendingAPDU else { return }
        pendingAPDU = nil
        guard data.count >= 3, data[0] == bleApduTag else {
            pendingAPDU = pending
            return
        }
        let packetIdx = (UInt16(data[1]) << 8) | UInt16(data[2])
        var offset = 3
        if packetIdx == 0 {
            guard data.count >= 5 else { pendingAPDU = pending; return }
            pending.totalLength = (Int(data[3]) << 8) | Int(data[4])
            offset = 5
        }
        pending.buffer.append(data.suffix(from: offset))
        if pending.totalLength > 0, pending.buffer.count >= pending.totalLength {
            let final = Data(pending.buffer.prefix(pending.totalLength))
            pending.continuation?.resume(returning: final)
            pending.continuation = nil
        } else {
            pendingAPDU = pending
        }
    }
}

File: Examples/android-test/app/src/main/java/com/benjaminchodroff/ledgerbtcexample/BLETransport.kt
// BLETransport: Android CoreBluetooth equivalent for the Ledger
// Nano X. Mechanical port of macOS/iOS BLETransport.swift.
//
// Caller must hold BLUETOOTH_CONNECT + BLUETOOTH_SCAN permissions
// (Android 12+) before calling `connect()`. The app must also
// pair the Ledger Nano X via Settings > Bluetooth FIRST; this
// transport reuses the OS-level pairing rather than initiating
// its own.
//
// UNTESTED on real hardware (Week 4 part B was implemented
// without an Android device available). Mechanical correctness
// only: shapes the API exactly the same way iOS/macOS does, with
// identical 153-byte MTU cap, 5-byte BLE framing, and 400/500ms
// Battery Service heartbeat. Real-device validation is pending.

package com.benjaminchodroff.ledgerbtcexample

import android.Manifest
import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattService
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.os.Build
import android.util.Log
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeoutOrNull
import uniffi.ledger_btc_core.ExchangeResponse
import uniffi.ledger_btc_core.Transport
import uniffi.ledger_btc_core.TransportException
import java.util.UUID

private const val TAG = "LedgerBLE"
private val LEDGER_SERVICE_UUID = UUID.fromString("13d63400-2c97-0004-0000-4c6564676572")
private val LEDGER_WRITE_UUID   = UUID.fromString("13d63400-2c97-0004-0002-4c6564676572")
private val LEDGER_NOTIFY_UUID  = UUID.fromString("13d63400-2c97-0004-0001-4c6564676572")
private val BATTERY_SERVICE_UUID = UUID.fromString("0000180f-0000-1000-8000-00805f9b34fb")
private val BATTERY_LEVEL_UUID   = UUID.fromString("00002a19-0000-1000-8000-00805f9b34fb")
private val CCC_DESCRIPTOR_UUID  = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")

private const val BLE_APDU_TAG: Byte = 0x05
private const val SAFE_MTU: Int = 153  // Verified Ledger Nano X cap.

/**
 * Production BLE transport. Untested without hardware; use
 * [MockTransport] for emulator UI work.
 *
 * Lifecycle: `connect()` → any number of `exchange(apdu)` calls →
 * the transport stays connected until process death or device
 * disconnect.
 */
@OptIn(kotlin.ExperimentalUnsignedTypes::class)
@SuppressLint("MissingPermission")  // caller is responsible for runtime perms
class BLETransport(
    private val context: Context,
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO),
    private val scanTimeoutMs: Long = 25_000,
    private val connectTimeoutMs: Long = 15_000,
    private val apduTimeoutMs: Long = 60_000,
) : Transport {

    private val bluetoothManager: BluetoothManager =
        context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
    private val bluetoothAdapter: BluetoothAdapter? = bluetoothManager.adapter

    private var gatt: BluetoothGatt? = null
    private var writeChar: BluetoothGattCharacteristic? = null
    private var notifyChar: BluetoothGattCharacteristic? = null
    private var batteryLevelChar: BluetoothGattCharacteristic? = null

    private var pendingAPDU: PendingAPDU? = null
    private class PendingAPDU(
        var totalLength: Int = -1,
        var buffer: ByteArray = ByteArray(0),
        val result: CompletableDeferred<ByteArray> = CompletableDeferred(),
    )

    private var heartbeatJob: Job? = null

    suspend fun connect() {
        if (bluetoothAdapter?.isEnabled != true) {
            throw TransportException.Io(reason = "Bluetooth is off or unsupported")
        }
        if (gatt != null) return

        val device = scanForLedger()
            ?: throw TransportException.Io(reason = "Ledger BLE device not found within ${scanTimeoutMs}ms")
        val connected = connectGatt(device)
        if (!connected) {
            throw TransportException.Disconnected(reason = "GATT connect failed")
        }
        discoverServices()
        if (writeChar == null || notifyChar == null) {
            throw TransportException.Io(reason = "Ledger APDU characteristics not found")
        }
    }

    fun disconnect() {
        heartbeatJob?.cancel()
        heartbeatJob = null
        gatt?.disconnect()
        gatt?.close()
        gatt = null
        writeChar = null
        notifyChar = null
        batteryLevelChar = null
    }

    // MARK: -- Transport

    override suspend fun exchange(apdu: ByteArray): ExchangeResponse {
        val writeCh = writeChar ?: throw TransportException.Disconnected(reason = "not connected")
        val gattRef = gatt ?: throw TransportException.Disconnected(reason = "not connected")
        // Fail any stale pending APDU before installing a new one.
        pendingAPDU?.result?.completeExceptionally(
            TransportException.Io(reason = "previous exchange abandoned")
        )
        pendingAPDU = null

        val pending = PendingAPDU()
        pendingAPDU = pending

        // Keep-alive heartbeat on the Battery Service. 400ms initial
        // delay so fast back-to-back rounds skip it; 500ms loop for
        // the long user-confirmation window. Matches verified-
        // stable iOS/macOS values.
        val battery = batteryLevelChar
        heartbeatJob = scope.launch {
            try {
                delay(400)
                while (true) {
                    val ch = battery ?: break
                    gattRef.readCharacteristic(ch)
                    delay(500)
                }
            } catch (_: Throwable) { /* cancellation */ }
        }

        // Send chunked APDU packets.
        val negotiatedMtu = SAFE_MTU  // we don't request MTU upgrade; 153 is the safe cap regardless
        val packets = framedPackets(apdu, negotiatedMtu)
        for (p in packets) {
            writeCh.value = p
            val ok = gattRef.writeCharacteristic(writeCh)
            if (!ok) {
                heartbeatJob?.cancel()
                throw TransportException.Io(reason = "writeCharacteristic returned false")
            }
        }

        val raw = withTimeoutOrNull(apduTimeoutMs) { pending.result.await() }
        heartbeatJob?.cancel()
        heartbeatJob = null
        if (raw == null) {
            pendingAPDU = null
            throw TransportException.Timeout(reason = "APDU response after ${apduTimeoutMs}ms")
        }
        if (raw.size < 2) {
            throw TransportException.Io(reason = "APDU response too short (${raw.size} bytes)")
        }
        val sw = (raw[raw.size - 2].toInt().and(0xff) shl 8) or raw[raw.size - 1].toInt().and(0xff)
        val data = raw.copyOfRange(0, raw.size - 2)
        return ExchangeResponse(statusWord = sw.toUShort(), data = data)
    }

    // MARK: -- internals

    private suspend fun scanForLedger(): BluetoothDevice? {
        val scanner = bluetoothAdapter?.bluetoothLeScanner
            ?: return null
        val found = CompletableDeferred<BluetoothDevice>()
        val callback = object : ScanCallback() {
            override fun onScanResult(callbackType: Int, result: ScanResult) {
                if (!found.isCompleted) found.complete(result.device)
            }
            override fun onScanFailed(errorCode: Int) {
                if (!found.isCompleted) {
                    found.completeExceptionally(
                        TransportException.Io(reason = "BLE scan failed (error $errorCode)")
                    )
                }
            }
        }
        val filters = listOf(
            ScanFilter.Builder().setServiceUuid(android.os.ParcelUuid(LEDGER_SERVICE_UUID)).build()
        )
        val settings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
            .build()
        scanner.startScan(filters, settings, callback)
        val device = try {
            withTimeoutOrNull(scanTimeoutMs) { found.await() }
        } finally {
            scanner.stopScan(callback)
        }
        return device
    }

    private suspend fun connectGatt(device: BluetoothDevice): Boolean {
        val ready = CompletableDeferred<Boolean>()
        val callback = object : BluetoothGattCallback() {
            override fun onConnectionStateChange(g: BluetoothGatt, status: Int, newState: Int) {
                if (newState == BluetoothProfile.STATE_CONNECTED) {
                    if (!ready.isCompleted) ready.complete(true)
                } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                    handleDisconnect(reason = "STATE_DISCONNECTED status=$status")
                    if (!ready.isCompleted) ready.complete(false)
                }
            }
            override fun onServicesDiscovered(g: BluetoothGatt, status: Int) {
                onServicesDiscoveredImpl(g, status)
            }
            override fun onCharacteristicChanged(g: BluetoothGatt, characteristic: BluetoothGattCharacteristic) {
                onCharacteristicChangedImpl(characteristic)
            }
            override fun onCharacteristicWrite(g: BluetoothGatt, characteristic: BluetoothGattCharacteristic, status: Int) {
                // No-op: we don't gate sequential writes on this; the
                // device tolerates back-to-back writeCharacteristic
                // calls so long as we don't exceed MTU per write.
            }
            override fun onCharacteristicRead(g: BluetoothGatt, characteristic: BluetoothGattCharacteristic, status: Int) {
                // Battery reads are heartbeat-only; nothing to do.
            }
        }
        val newGatt = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            device.connectGatt(context, false, callback, BluetoothDevice.TRANSPORT_LE)
        } else {
            device.connectGatt(context, false, callback)
        }
        this.gatt = newGatt
        val connected = withTimeoutOrNull(connectTimeoutMs) { ready.await() } ?: false
        return connected
    }

    private suspend fun discoverServices() {
        val g = gatt ?: return
        val discovered = CompletableDeferred<Boolean>()
        servicesDiscoveryGate = discovered
        g.discoverServices()
        withTimeoutOrNull(connectTimeoutMs) { discovered.await() }
        servicesDiscoveryGate = null
    }

    private var servicesDiscoveryGate: CompletableDeferred<Boolean>? = null

    private fun onServicesDiscoveredImpl(g: BluetoothGatt, status: Int) {
        if (status != BluetoothGatt.GATT_SUCCESS) {
            servicesDiscoveryGate?.complete(false)
            return
        }
        val ledgerSvc: BluetoothGattService? = g.getService(LEDGER_SERVICE_UUID)
        val batterySvc: BluetoothGattService? = g.getService(BATTERY_SERVICE_UUID)
        writeChar = ledgerSvc?.getCharacteristic(LEDGER_WRITE_UUID)
        notifyChar = ledgerSvc?.getCharacteristic(LEDGER_NOTIFY_UUID)
        batteryLevelChar = batterySvc?.getCharacteristic(BATTERY_LEVEL_UUID)
        val nc = notifyChar
        if (nc != null) {
            g.setCharacteristicNotification(nc, true)
            // Required to actually enable notifications on most BLE
            // stacks: write 0x01 0x00 to the CCC descriptor.
            val ccc = nc.getDescriptor(CCC_DESCRIPTOR_UUID)
            if (ccc != null) {
                ccc.value = BluetoothGattDescriptorEnable
                g.writeDescriptor(ccc)
            }
        }
        servicesDiscoveryGate?.complete(true)
    }

    private fun onCharacteristicChangedImpl(characteristic: BluetoothGattCharacteristic) {
        // Filter: only the Ledger notify characteristic carries
        // APDU response packets. Battery reads come through
        // onCharacteristicRead and we ignore them.
        if (characteristic.uuid != LEDGER_NOTIFY_UUID) return
        val data = characteristic.value ?: return
        val pending = pendingAPDU ?: return
        if (data.size < 3 || data[0] != BLE_APDU_TAG) return
        val packetIdx = (data[1].toInt().and(0xff) shl 8) or data[2].toInt().and(0xff)
        var offset = 3
        if (packetIdx == 0) {
            if (data.size < 5) return
            pending.totalLength = (data[3].toInt().and(0xff) shl 8) or data[4].toInt().and(0xff)
            offset = 5
        }
        pending.buffer += data.copyOfRange(offset, data.size)
        if (pending.totalLength > 0 && pending.buffer.size >= pending.totalLength) {
            pending.result.complete(pending.buffer.copyOf(pending.totalLength))
            pendingAPDU = null
        }
    }

    private fun handleDisconnect(reason: String) {
        val err = TransportException.Disconnected(reason = reason)
        pendingAPDU?.result?.completeExceptionally(err)
        pendingAPDU = null
        heartbeatJob?.cancel()
        heartbeatJob = null
    }
}

// CCC descriptor bytes that enable BLE notifications on the
// characteristic it belongs to. Same value across all Android
// versions; defined as a constant to keep the discovery path tidy.
private val BluetoothGattDescriptorEnable = byteArrayOf(0x01, 0x00)

private fun framedPackets(apdu: ByteArray, mtu: Int): List<ByteArray> {
    val out = ArrayList<ByteArray>()
    var remaining = apdu
    var idx = 0
    while (remaining.isNotEmpty()) {
        val header = ArrayList<Byte>()
        header.add(BLE_APDU_TAG)
        header.add(((idx shr 8) and 0xff).toByte())
        header.add((idx and 0xff).toByte())
        if (idx == 0) {
            header.add(((apdu.size shr 8) and 0xff).toByte())
            header.add((apdu.size and 0xff).toByte())
        }
        var take = mtu - header.size
        if (take > remaining.size) take = remaining.size
        val chunk = ByteArray(header.size + take)
        for ((i, b) in header.withIndex()) chunk[i] = b
        System.arraycopy(remaining, 0, chunk, header.size, take)
        out.add(chunk)
        remaining = remaining.copyOfRange(take, remaining.size)
        idx += 1
    }
    return out
}
